library/pem.c

```diff
@@ -431,21 +453,36 @@ int mbedtls_pem_read_buffer(mbedtls_pem_context *ctx, const char *header, const
             return ret;
         }
 
+        /* Check PKCS padding and update data length based on padding info.
+         * This can be used to detect invalid padding data and password
+         * mismatches. */
+        ret = pem_check_pkcs_padding(buf, len, &len);
+        if (ret != 0) {
+            mbedtls_free(buf);
+            return ret;
+        }
+
         /*
-         * The result will be ASN.1 starting with a SEQUENCE tag. Parse it
-         * with ASN.1 functions in order to:
-         * - Have an heuristic guess about password mismatches.
-         * - Update len variable to the amount of valid data inside buf.
+         * In RFC1421 PEM is used as container for DER (ASN.1) content so we
```

Comment:
Non-blocker: now that we're checking the padding, the old reasons to parse the ASN.1 header don't apply anymore:

> - Have an heuristic guess about password mismatches. → we get that from the padding
> - Update len variable to the amount of valid data inside buf. → we get that from the padding

So maybe we should remove the ASN.1 call to reduce complexity? Both here, and we wouldn't need to [add yet another condition for including ASN.1 functions](https://github.com/Mbed-TLS/mbedtls/pull/8804/commits/b1f6d2ad6fa2621772d35bda1835d8b0e1fa1d02).